reframe/core/schedulers/sge.py

```diff
@@ -0,0 +1,225 @@
+
+        job._jobid = jobid_match.group('jobid')
+        job._submit_time = time.time()
+
+    def wait(self, job):
+        intervals = itertools.cycle([1, 2, 3])
+        while not self.finished(job):
+            self.poll(job)
+            time.sleep(next(intervals))
+
+    def cancel(self, job):
+        time_from_submit = time.time() - job.submit_time
+        if time_from_submit < SGE_CANCEL_DELAY:
+            time.sleep(SGE_CANCEL_DELAY - time_from_submit)
+
+        _run_strict(f'qdel {job.jobid}', timeout=self._submit_timeout)
+        job._cancelled = True
+
+    def finished(self, job):
+        if job.exception:
+            raise job.exception
+
+        return job.completed
+
+    def poll(self, *jobs):
+        if jobs:
+            # Filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        user = osext.osuser()
+        completed = osext.run_command(
+            f'qstat -xml -u {user}'
+        )
+
+        if completed.returncode != 0:
+            raise JobSchedulerError(
+                f'qstat failed with exit code {completed.returncode} '
+                f'(standard error follows):\n{completed.stderr}'
+            )
+
+        root = ET.fromstring(completed.stdout)
+
+        # Store information for each job separately
+        jobinfo = {}
+        for queue_info in root:
+
+            # Reads the XML and prints jobs with status belonging to user.
+            if queue_info is None:
+                raise JobSchedulerError('Decomposition error!\n')
+
+            for job_list in queue_info:
+                if job_list.find("JB_owner").text != user:
+                    # Not a job of this user.
+                    continue
+
+                job_number = job_list.find("JB_job_number").text
+
+                if job_number not in [job.jobid for job in jobs]:
+                    # Not a reframe job.
+                    continue
+
+                state = job_list.find("state").text
+
+                # For the list of known statuses see `man 5 sge_status`
+                # (https://arc.liv.ac.uk/SGE/htmlman/htmlman5/sge_status.html)
+                if state in ['r', 'hr', 't', 'Rr', 'Rt']:
+                    jobinfo[job_number] = 'RUNNING'
+                elif state in ['qw', 'Rq', 'hqw', 'hRwq']:
+                    jobinfo[job_number] = 'PENDING'
+                elif state in ['s', 'ts', 'S', 'tS', 'T', 'tT', 'Rs',
+                               'Rts', 'RS', 'RtS', 'RT', 'RtT']:
+                    jobinfo[job_number] = 'SUSPENDED'
+                elif state in ['Eqw', 'Ehqw', 'EhRqw']:
+                    jobinfo[job_number] = 'ERROR'
+                elif state in ['dr', 'dt', 'dRr', 'dRt', 'ds',
+                               'dS', 'dT', 'dRs', 'dRS', 'dRT']:
+                    jobinfo[job_number] = 'DELETING'
+                elif state == 'z':
+                    jobinfo[job_number] = 'COMPLETED'
```

Comment:
You can make a mapping similar to the `JOB_STATES` variable and set directly the `job._state`.

Reply:
Sorry, my bad. I had a look again at the code and you don't have immediate access to the job in this part. You should leave it as it is.